src/compiler/index.js

```diff
@@ -43,31 +46,76 @@ export default class Compiler {
 
         code = stripBom(code).toString();
 
-        const compiler = find(testFileCompilers, c => c.canCompile(code, filename));
+        const compilerInfo = find(testFileCompilersInfo, ([/* name */, compiler]) => compiler.canCompile(code, filename));
+
+        if (!compilerInfo)
+            return null;
+
+        const [compilerName, compiler] = compilerInfo;
+
+        return {
+            filename,
+            code,
+            compiler,
+            compilerName,
 
-        return compiler ? await compiler.compile(code, filename) : null;
+            compiledCode: null
+        };
     }
 
-    async getTests () {
-        const sourceChunks = chunk(this.sources, SOURCE_CHUNK_LENGTH);
-        let tests        = [];
-        let compileUnits = [];
+    async _createTestFilesInfo (filenames) {
+        const testFilesInfo = await Promise.all(filenames.map(filename => this._createTestFileInfo(filename)));
+
+        return testFilesInfo.filter(info => !!info);
+    }
+
+    async _precompileFiles (compiler, testFilesInfo) {
+        if (!compiler.canPrecompile)
+            return;
+
+        const precompiledCode = await compiler.precompile(testFilesInfo);
+
+        zipWith(testFilesInfo, precompiledCode.map(compiledCode => ({ compiledCode })), assign);
+    }
+
+    async _getTests ({ compiler, filename, code, compiledCode }) {
+        if (compiledCode)
+            return await compiler.execute(compiledCode, filename);
+
+        return await compiler.compile(code, filename);
+    }
 
+    async _compileTestFiles (filenames) {
+        const testFilesInfo = await this._createTestFilesInfo(filenames);
+        const compilerTasks = groupBy(testFilesInfo, ({ compilerName }) => compilerName);
+        const compilerNames = Object.keys(compilerTasks);
+
+        await Promise.all(compilerNames.map(compilerName => this._precompileFiles(testFileCompilersRegistry[compilerName], compilerTasks[compilerName])));
```

Comment:
we already have `testFilesInfo` here which contains `compiler`. Can we get rid of `testFileCompilersRegistry`?